src/group/libp2p_group_gossip_server.erl

```diff
@@ -377,6 +383,62 @@ assign_target(WorkerPid, WorkerRef, TargetAddrs, State=#state{workers=Workers, s
             end
     end.
 
+maybe_lookup_seed_in_dns(TargetAddrs) ->
+    case application:get_env(libp2p, use_dns_for_seeds, false) of
+        false ->
+            TargetAddrs;
+        true ->
+            lookup_seed_from_dns(TargetAddrs)
+    end.
+
+choose_random_element([E]) -> E;
+choose_random_element(L) when is_list(L) ->
+    lists:nth(rand:uniform(length(L), L)).
+
+%% We will (try to) lookup seed IPs from a DNS cname, and fall back to
+%% using the static seed list
+lookup_seed_from_dns(TargetAddrs) ->
+    case application:get_env(libp2p, seed_dns_cname, undefined) of
+        undefined ->
+            lager:error("Configured to use DNS to lookup seed node IP, but the cname is undefined", []),
+            TargetAddrs;
+        CName ->
+            case attempt_dns_lookup(CName, ?DNS_RETRIES) of
+                {error, _} = Error ->
+                    lager:error("DNS lookup of ~p resulted in ~p; falling back", [CName, Error]),
+                    TargetAddrs;
+                {ok, DNSRecord} ->
+                    lager:debug("successful DNS lookup result: ~p", [lager:pr(DNSRecord, inet)]),
+                    %% to help migitate possible eclipse attacks we will blend the DNS results
+                    %% with the static list of seed nodes
+                    convert_dns_records(DNSRecord) ++ TargetAddrs
+            end
+    end.
+
+attempt_dns_lookup(_Name, 0) -> {error, too_many_lookup_attempts};
+attempt_dns_lookup(Name, Attempts) ->
+    case inet_res:gethostbyname(Name, inet, ?DNS_TIMEOUT) of
+        {error, _} = Error ->
+            lager:debug("attempt ~p: got ~p", [Attempts, Error]),
+            timer:sleep(?DNS_SLEEP),
+            attempt_dns_lookup(Name, Attempts - 1);
+        {ok, HostRec} -> {ok, HostRec}
+    end.
+
+convert_dns_records(#hostent{h_addr_list = AddrList}) ->
+    [ format_dns_addr(A) || A <- AddrList ].
+
+format_dns_addr(Addr) ->
+    IPStr = inet:ntoa(Addr),
+    PortList = application:get_env(libp2p, seed_dns_ports, ?DEFAULT_SEED_DNS_PORTS),
+    PortStr = to_string(choose_random_element(PortList)),
+    %% XXX this can potentially return `error(badarg)' so maybe try/catch?
+    libp2p_crypto:p2p_to_pubkey_bin("/ip4/" ++ IPStr ++ "/tcp/" ++ PortStr).
```

Comment:
This is not what this function is for, you should just return the concatenated string here?